internal/Yesod/Session/Memcache/Storage.hs

```diff
@@ -0,0 +1,98 @@
+module Yesod.Session.Memcache.Storage
+  ( memcacheStorage
+  , SessionPersistence (..)
+  ) where
+
+import Internal.Prelude
+
+import Database.Memcache.Client qualified as Memcache
+import Database.Memcache.Types qualified as Memcache
+import Session.Key
+import Session.Timing.Math (nextExpires)
+import Session.Timing.Options (TimingOptions (timeout))
+import Session.Timing.Time (Time (..))
+import Time (UTCTime)
+import Yesod.Core (SessionMap)
+import Yesod.Session.Memcache.Expiration
+  ( MemcacheExpiration
+  , fromUTC
+  , noExpiration
+  )
+import Yesod.Session.Options (Options (timing))
+import Yesod.Session.SessionType
+import Yesod.Session.Storage.Exceptions
+import Yesod.Session.Storage.Operation
+
+-- | Mapping between 'Session' and Memcache representation.
+data SessionPersistence = SessionPersistence
+  { databaseKey :: SessionKey -> Memcache.Key
+  , toDatabase :: (SessionMap, Time UTCTime) -> Memcache.Value
+  , fromDatabase
+      :: Memcache.Value
+      -> Either SomeException (SessionMap, Time UTCTime)
+  , client :: Memcache.Client
+  , expiration :: MemcacheExpiration
+  }
+
+memcacheStorage
+  :: forall m result
+   . (MonadThrow m, MonadIO m)
+  => SessionPersistence
+  -> Options IO IO
+  -> StorageOperation result
+  -> m result
+memcacheStorage sp opt = \case
+  GetSession sessionKey -> do
+    mValue <-
+      liftIO $ fmap fstOf3 <$> Memcache.get sp.client (sp.databaseKey sessionKey)
+
+    case mValue of
+      Nothing -> pure Nothing
+      Just value -> do
+        let mkSession = uncurry $ Session sessionKey
+
+        either throwM (pure . pure . mkSession) $ sp.fromDatabase value
+  DeleteSession sessionKey -> do
+    void $ liftIO $ Memcache.delete sp.client (sp.databaseKey sessionKey) bypassCAS
+  InsertSession session -> do
+    let
+      key = sp.databaseKey session.key
+      value = sp.toDatabase (session.map, session.time)
+
+    mVersion <-
+      liftIO
+        $ Memcache.add sp.client key value defaultFlags
+        $ getNextExpires opt.timing.timeout session.time
+    throwOnNothing SessionAlreadyExists mVersion
+  ReplaceSession session -> do
+    let key = sp.databaseKey session.key
+    mVersion <-
+      liftIO
+        $ Memcache.replace
+          sp.client
+          key
+          (sp.toDatabase (session.map, session.time))
+          defaultFlags
+          (getNextExpires opt.timing.timeout session.time)
+          bypassCAS
+    throwOnNothing SessionDoesNotExist mVersion
+ where
+  throwOnNothing exception maybeValue = maybe (throwWithCallStack exception) (const $ pure ()) maybeValue
+  fstOf3 (a, _, _) = a
+  getNextExpires timeout time = fromMaybe noExpiration $ nextExpires timeout time >>= fromUTC
```

Comment:
In the extraordinary event that `fromUTC` fails, I think this should throw the exception rather than silently failing to set an expiration time.

```haskell
  getNextExpires timeout time =
    maybe (pure noExpiration) fromUTC $ nextExpires timeout time
```

Reply:
Great catch and quite subtle to me that `getNextExpires` was catching the error thrown from `nextExpires`. As a result, I added testing to `getNextExpires` to make sure it bubbles up the error.